exchanges/exchanges.go

```diff
@@ -2851,3 +2887,322 @@ func (poloniex *PoloniexExchange) Refresh() {
+		address: "wss://" + dcr.cfg.Host + "/ws",
+		tlsConfig: &tls.Config{
+			RootCAs:    pool,
+			ServerName: dcr.cfg.CertHost,
+		},
+	})
+	if err != nil {
+		dcr.setWsFail(fmt.Errorf("dcr.connectWs: %v", err))
+		return
+	}
+
+	err = dcr.wsSend(decredDEXOrderBookSubscription)
+	if err != nil {
+		dcr.setWsFail(fmt.Errorf("error sending order book subscription: %v", err))
+	}
+}
+
+// processWsMessage is DecredDEX's WebsocketProcessor. Handles messages of type
+// *msgjson.Message.
+func (dcr *DecredDEX) processWsMessage(raw []byte) {
+	msg, err := msgjson.DecodeMessage(raw)
+	if err != nil {
+		dcr.setWsFail(fmt.Errorf("DecodeMessage error: %v", err))
+		return
+	}
+
+	dcr.orderMtx.Lock()
+	defer dcr.orderMtx.Unlock()
+	dcr.stamp = time.Now().Unix()
+	switch msg.Type {
+	case msgjson.Response:
+		// The only request we make is for the order book subscription. The
+		// response will include the full order book.
+		if msg.ID != DexSubscriptionID {
+			log.Errorf("received response from %s with unknown message ID: %s", dcr.cfg.Host, string(raw))
+			return
+		}
+		ob := new(msgjson.OrderBook)
+		err := msg.UnmarshalResult(ob)
+		if err != nil {
+			dcr.setWsFail(fmt.Errorf("error unmarshaling orderbook response: %v", err))
+			return
+		}
+		dcr.setOrderBook(ob)
+	case msgjson.Notification:
+		switch msg.Route {
+		case msgjson.BookOrderRoute:
+			bookOrder := new(msgjson.BookOrderNote)
+			err := msg.Unmarshal(bookOrder)
+			if err != nil {
+				dcr.setWsFail(fmt.Errorf("book_order Unmarshal error: %v", err))
+				return
+			}
+			if !dcr.checkSeq(bookOrder.Seq) {
+				return
+			}
+			dcr.bookOrder(bookOrder)
+		case msgjson.UnbookOrderRoute:
+			unbookOrder := new(msgjson.UnbookOrderNote)
+			err := msg.Unmarshal(unbookOrder)
+			if err != nil {
+				dcr.setWsFail(fmt.Errorf("unbook_order Unmarshal error: %v", err))
+				return
+			}
+			if !dcr.checkSeq(unbookOrder.Seq) {
+				return
+			}
+			dcr.unbookOrder(unbookOrder)
+		case msgjson.UpdateRemainingRoute:
+			update := new(msgjson.UpdateRemainingNote)
+			err := msg.Unmarshal(update)
+			if err != nil {
+				dcr.setWsFail(fmt.Errorf("update_remaining Unmarshal error: %v", err))
+				return
+			}
+			if !dcr.checkSeq(update.Seq) {
+				return
+			}
+			dcr.updateRemaining(update)
+		case msgjson.EpochOrderRoute:
+			// We don't actually track epoch orders, but we need to progress the
+			// sequence.
+			note := new(msgjson.EpochOrderNote)
+			err := msg.Unmarshal(note)
+			if err != nil {
+				dcr.setWsFail(fmt.Errorf("epoch_order Unmarshal error: %v", err))
+				return
+			}
+			dcr.checkSeq(note.Seq)
+			return // Skip wsUpdate. Nothing has changed.
+		case msgjson.SuspensionRoute:
+			note := new(msgjson.TradeSuspension)
+			err := msg.Unmarshal(note)
+			if err != nil {
+				dcr.setWsFail(fmt.Errorf("suspension Unmarshal error: %v", err))
+				return
+			}
+			if !dcr.checkSeq(note.Seq) {
+				return
+			}
+			if note.SuspendTime > 0 || note.Persist {
```

Comment:
If `note.Persist == true`, `Seq` will be zero/unset, so I think you want to checkSeq after checking note.Persist